src/web/page/templates.rs

```diff
@@ -0,0 +1,191 @@
+use super::TEMPLATE_DATA;
+use crate::error::Result;
+use arc_swap::ArcSwap;
+use serde_json::Value;
+use std::collections::HashMap;
+use tera::{Result as TeraResult, Tera};
+
+/// Holds all data relevant to templating
+pub(crate) struct TemplateData {
+    /// The actual templates, stored in an `ArcSwap` so that they're hot-swappable
+    // TODO: Conditional compilation so it's not always wrapped, the `ArcSwap` is unneeded overhead for prod
+    pub templates: ArcSwap<Tera>,
+    /// The current global alert, serialized into a json value
+    global_alert: Value,
+    /// The version of docs.rs, serialized into a json value
+    docsrs_version: Value,
+    /// The current resource suffix of rustc, serialized into a json value
+    resource_suffix: Value,
+}
+
+impl TemplateData {
+    pub fn new() -> Result<Self> {
+        log::trace!("Loading templates");
+
+        let data = Self {
+            templates: ArcSwap::from_pointee(load_templates()?),
+            global_alert: serde_json::to_value(crate::GLOBAL_ALERT)?,
+            docsrs_version: Value::String(crate::BUILD_VERSION.to_owned()),
+            resource_suffix: Value::String(load_rustc_resource_suffix().unwrap_or_else(|err| {
+                log::error!("Failed to load rustc resource suffix: {:?}", err);
+                String::from("???")
+            })),
+        };
+
+        log::trace!("Finished loading templates");
+
+        Ok(data)
+    }
+
+    pub fn start_template_reloading() {
+        use std::{sync::Arc, thread, time::Duration};
+
+        thread::spawn(|| loop {
+            match load_templates() {
+                Ok(templates) => {
+                    log::info!("Reloaded templates");
+                    TEMPLATE_DATA.templates.swap(Arc::new(templates));
+                    thread::sleep(Duration::from_secs(10));
```

Comment:
So, the whole reload logic is just reloading everything every 10 seconds? I'm not sure that's the best approach, especially because it takes less than 10 seconds to restart the docs.rs application if it's running outside of docker, so this just makes the experience worse.

I'd expect template reloading to watch for changes in the templates directory, and reload once something is changed. I understand that's more complex to implement, but we can always back it out of this PR and land it in a followup.

Reply:
It's definitely not ideal in any sense, but that was my last-ditch effort. I tried to use `notify` (which is supposed to be a cross-platform file/dir watcher), but it just never worked for me. No matter what I did, the changes weren't recognized and so things never reloaded. I can try again to get notify to work, but if all else fails this isn't the *worst* solution ever, since reloading is a development thing